sdks/python/apache_beam/runners/common.py

```diff
@@ -214,62 +254,125 @@ def _dofn_per_window_invoker(self, element):
-    # otherwise as none of the arguments are changing
-    if self.has_windowed_inputs and len(element.windows) != 1:
-      for w in element.windows:
-        self._dofn_per_window_invoker(
-            WindowedValue(element.value, element.timestamp, (w,)))
+
+class DoFnRunner(Receiver):
+  """A helper class for executing ParDo operations.
+  """
+
+  def __init__(self,
+               fn,
+               args,
+               kwargs,
+               side_inputs,
+               windowing,
+               context=None,
+               tagged_receivers=None,
+               logger=None,
+               step_name=None,
+               # Preferred alternative to logger
+               # TODO(robertwb): Remove once all runners are updated.
+               logging_context=None,
+               # Preferred alternative to context
+               # TODO(robertwb): Remove once all runners are updated.
+               state=None,
+               scoped_metrics_container=None):
+    """Initializes a DoFnRunner.
+
+    Args:
+      fn: user DoFn to invoke
+      args: positional side input arguments (static and placeholder), if any
+      kwargs: keyword side input arguments (static and placeholder), if any
+      side_inputs: list of sideinput.SideInputMaps for deferred side inputs
+      windowing: windowing properties of the output PCollection(s)
+      context: a DoFnContext to use (deprecated)
+      tagged_receivers: a dict of tag name to Receiver objects
+      logger: a logging module (deprecated)
+      step_name: the name of this step
+      logging_context: a LoggingContext object
+      state: handle for accessing DoFn state
+      scoped_metrics_container: Context switcher for metrics container
+    """
+    self.tagged_receivers = tagged_receivers
+    self.scoped_metrics_container = (scoped_metrics_container
+                                     or ScopedMetricsContainer())
+
+    # Optimize for the common case.
+    self.main_receivers = as_receiver(tagged_receivers[None])
+
+    self.step_name = step_name
+
+    # Need to support multiple iterations.
+    side_inputs = list(side_inputs)
+
+    if logging_context:
+      self.logging_context = logging_context
     else:
-      self._dofn_per_window_invoker(element)
+      self.logging_context = get_logging_context(logger, step_name=step_name)
+
+    # TODO(sourabh): Deprecate the use of context
+    if state:
+      assert context is None
+      context = DoFnContext(step_name, state=state)
+    else:
+      assert context is not None
+      context = context
+
+    self.context = context
+
+    do_fn_signature = DoFnSignature.create_signature(fn)
+
+    default_arg_values = do_fn_signature.process_method.defaults
+    use_simple_invoker = (
+        not side_inputs and not args and not kwargs and not default_arg_values)
 
-  def _invoke_bundle_method(self, method):
+    self.window_fn = windowing.windowfn
+
+    self.do_fn_invoker = DoFnInvoker.create_invoker(
```

Comment:
we should move the use_simple_invoker boolean inside this as that makes it a pure factory .. the runner shouldn't decide what invoker is returned.